drivers/platform/chibios/chibios_spi.c

```diff
@@ -0,0 +1,145 @@
+ *    contributors may be used to endorse or promote products derived
+ *    from this software without specific prior written permission.
+ *  - The use of this software may or may not infringe the patent rights
+ *    of one or more patent holders.  This license does not release you
+ *    from the requirement that you obtain separate licenses from these
+ *    patent holders to use this software.
+ *  - Use of the software either in source or binary form, must be run
+ *    on or directly connected to an Analog Devices Inc. component.
+ *
+ * THIS SOFTWARE IS PROVIDED BY ANALOG DEVICES "AS IS" AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, NON-INFRINGEMENT,
+ * MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL ANALOG DEVICES BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, INTELLECTUAL PROPERTY RIGHTS, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+ * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*******************************************************************************/
+#include <stdlib.h>
+#include <errno.h>
+#include "no_os_util.h"
+#include "no_os_spi.h"
+#include "no_os_alloc.h"
+#include "chibios_spi.h"
+
+#if (HAL_USE_SPI==TRUE)
+
+/**
+ * @brief Initialize the SPI communication peripheral.
+ * @param desc - The SPI descriptor.
+ * @param param - The structure that contains the SPI parameters.
+ * @return 0 in case of success, -1 otherwise.
+ */
+int32_t chibios_spi_init(struct no_os_spi_desc **desc,
+			 const struct no_os_spi_init_param *param)
+{
+	int32_t ret;
+
+	struct no_os_spi_desc *spi_desc;
+	struct chibios_spi_desc *sdesc;
+
+	if (!desc || !param)
+		return -EINVAL;
+
+	spi_desc = (struct no_os_spi_desc *)no_os_malloc(sizeof(struct no_os_spi_desc));
+
+	if (!spi_desc) {
+		ret = -ENOMEM;
+		goto error_1;
+	}
+
+	sdesc = (struct chibios_spi_desc*)no_os_malloc(sizeof(struct chibios_spi_desc));
+
+	if (!sdesc) {
+		ret = -ENOMEM;
+		goto error_2;
+	}
+
+	sdesc->hspi = ((struct chibios_spi_init_param *)(param->extra))->hspi;
+	sdesc->spicfg = ((struct chibios_spi_init_param *)(param->extra))->spicfg;
+	spi_desc->extra = sdesc;
+
+	spiStart((sdesc->hspi), (sdesc->spicfg));
+
+	*desc = spi_desc;
+	return 0;
+error_1:
+	no_os_free(spi_desc);
+error_2:
+	no_os_free(sdesc);
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by no_os_spi_init().
+ * @param desc - The SPI descriptor.
+ * @return 0 in case of success, -1 otherwise.
+ */
+int32_t chibios_spi_remove(struct no_os_spi_desc *desc)
+{
+	struct chibios_spi_desc *sdesc = desc->extra;
+	spiStop((sdesc->hspi));
```

Comment:
Might be a good idea to null check `desc`. Also you may set the `desc` variable to `NULL` after you free it to avoid a double free.